runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.core;
+
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.SetCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.ReadyCheckingSideInputReader;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.state.AccumulatorCombiningState;
+import org.apache.beam.sdk.util.state.StateInternals;
+import org.apache.beam.sdk.util.state.StateNamespaces;
+import org.apache.beam.sdk.util.state.StateTag;
+import org.apache.beam.sdk.util.state.StateTags;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.values.PCollectionView;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Generic side input handler that uses {@link StateInternals} to store all data. Both the actual
+ * side-input data and data about the windows for which we have side inputs available are stored
+ * using {@code StateInternals}.
+ *
+ * <p>The given {@code StateInternals} must not be scoped to an element key. The state
+ * must instead be scoped to one key group for which the side input is being managed.
+ *
+ * <p>This is useful for runners that transmit the side-input elements in band, as opposed
+ * to how Dataflow has an external service for managing side inputs.
+ *
+ * <p>Note: storing the available windows in an extra state is redundant for now but in the
+ * future we might want to know which windows we have available so that we can garbage collect
+ * side input data. For now, this will never clean up side-input data because we have no way
+ * of knowing when we reach the GC horizon.
+ */
+public class SideInputHandler implements ReadyCheckingSideInputReader {
+
+  /** The list of side inputs that we're handling. */
+  protected final Collection<PCollectionView<?>> sideInputs;
+
+  /** State internals that are scoped not to the key of a value but instead to one key group. */
```

Comment:
Another way of describing this is that side inputs are, rather by definition, global data. So there's no relevant key.